File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/retrieval/model/RetrievalDocument.java
package com.study.webflux.rag.domain.retrieval.model;

import java.util.Map;

public record RetrievalDocument(
	String content,
	com.study.webflux.rag.domain.retrieval.model.SimilarityScore score,
	Map<String, Object> metadata
) {
	public RetrievalDocument {
		if (content == null || content.isBlank()) {
			throw new IllegalArgumentException("content cannot be null or blank");
		}
		if (score == null) {
			throw new IllegalArgumentException("score cannot be null");
		}
		if (metadata == null) {
			metadata = Map.of();
		}
	}

	public static RetrievalDocument of(String content, int score) {
		return new RetrievalDocument(content,
			com.study.webflux.rag.domain.retrieval.model.SimilarityScore.of(score), Map.of());
	}

	public static RetrievalDocument withMetadata(String content,
		int score,
		Map<String, Object> metadata) {
		return new RetrievalDocument(content,
			com.study.webflux.rag.domain.retrieval.model.SimilarityScore.of(score), metadata);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/retrieval/model/RetrievalDocument.kt
package com.miyou.app.domain.retrieval.model

data class RetrievalDocument(
    val content: String,
    val score: SimilarityScore,
    val metadata: Map<String, Any> = emptyMap(),
) {
    init {
        require(content.isNotBlank()) { "content cannot be null or blank" }
    }

    companion object {
        @JvmStatic
        fun of(
            content: String,
            score: Int,
        ): RetrievalDocument = RetrievalDocument(content, SimilarityScore.of(score), emptyMap())

        @JvmStatic
        fun withMetadata(
            content: String,
            score: Int,
            metadata: Map<String, Any>?,
        ): RetrievalDocument = RetrievalDocument(content, SimilarityScore.of(score), metadata.orEmpty())
    }

    fun content(): String = content

    fun score(): SimilarityScore = score

    fun metadata(): Map<String, Any> = metadata
}
